Lorawan rtc

FRASTM opened the pull request:
stm32 BSP LoRaWan with RTC instance

FRASTM committed:
stm32 BSP LoRaWan with RTC instance of the stm32WL

The Alarm B of the RTC is used for LoRaWan purpose.
Use the LSE as clock source for the RTC instance.
set the free running mix mode.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
stm32duino LoraWan enabling the RTC alarm B

Mask is set to ALL in case of MIX mode, so that interrupt is trigged
on the sub-second basis (other calendar values do not care)

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
STM32LoRaWan set the setRTCMode before the rtc.begin to set MIX mode

The RTC.begin does not propose the RTCmode parameter.
This is configured with the setRTCMode method before .begin
The RTC irq (for Alarm B) is handled directly by the STM32RTC library

Signed-off-by: Francois Ramu <francois.ramu@st.com>

fpistm commented on src/BSP/timer_if.c:
```diff
@@ -223,22 +253,15 @@ UTIL_TIMER_Status_t TIMER_IF_StartTimer(uint32_t timeout)
-  }
+
+  /* Program ALARM A on subsecond, mask is 32 (and fixed to RTC_ALARMMASK_NONE for calendar) */
+  RTC_StartAlarm(RTC_ALARM_A, 0, 0, 0, 0, UINT32_MAX - timeout, RTC_HOURFORMAT12_PM, 32UL);
```
Use RTC_ALARM_B instead to reserve RTC_ALARM_A for other usage.

fpistm commented on src/BSP/timer_if.c:
```diff
@@ -251,10 +274,15 @@ UTIL_TIMER_Status_t TIMER_IF_StopTimer(void)
   /* Clear RTC Alarm Flag */
   __HAL_RTC_ALARM_CLEAR_FLAG(&hrtc, RTC_FLAG_ALRAF);
   /* Disable the Alarm A interrupt */
   HAL_RTC_DeactivateAlarm(&hrtc, RTC_ALARM_A);
```
ditto
RTC_ALARM_B

fpistm commented on src/BSP/timer_if.c:
```diff
@@ -251,10 +274,15 @@ UTIL_TIMER_Status_t TIMER_IF_StopTimer(void)
   /* Disable the Alarm A interrupt */
   HAL_RTC_DeactivateAlarm(&hrtc, RTC_ALARM_A);
+#else
+  rtc.disableAlarm(RTC_ALARM_A);
```
ditto 
RTC_ALARM_B

fpistm commented on src/BSP/timer_if.c:
```diff
@@ -223,22 +253,15 @@ UTIL_TIMER_Status_t TIMER_IF_StartTimer(uint32_t timeout)
-  }
+
+  /* Program ALARM A on subsecond, mask is 32 (and fixed to RTC_ALARMMASK_NONE for calendar) */
+  RTC_StartAlarm(RTC_ALARM_A, 0, 0, 0, 0, UINT32_MAX - timeout, RTC_HOURFORMAT12_PM, 32UL);
```
why RTC_HOURFORMAT12_PM ? 
below it is set to 
rtc.begin(true, STM32RTC::HOUR_24);

FRASTM replied:
yes, the parameter in the RTC_StartAlarm is of type `hourAM_PM_t` (HOUR_AM or  HOUR_PM) which will be ignored in the function :  ` /* Ignore time AM PM configuration if in 24 hours format */`

FRASTM replied:
Use the RTC handle as a global from the STM32_RTC library 
Use Alarm B of the RTC to trig on SubSeconds

FRASTM committed:
simplify the BSP to keep are timer_f and include the RTC library

Include the STM32RTC library to all the BSP as only few functions
remain in the BSP as the RTC is instantiated by the STM32RTC library

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Using the **ALARM B** for the LoraWan and keep Alarm A for other purpose
Remove `#if 0/1` code  sections

FRASTM committed:
configure and set the AlarmB to use in MIX mode
Directly call the RTC functions from the STM32duino RTC library
Let the IsEnabled.RtcFeatures be handled by the RTC library
